Messaging v1beta1 Types

vaikas opened the pull request:
Addresses #1869

## Proposed Changes

- Add v1beta1.[subscription,channel,inmemorychannel]
- Add corresponding duckv1beta1 types

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
NONE (missing reconcilers, etc. just the types).
```

n3wscott commented on pkg/apis/messaging/v1beta1/subscription_types.go:
```diff
@@ -0,0 +1,147 @@
+// Subscription.
+type SubscriptionStatusPhysicalSubscription struct {
+	// SubscriberURI is the fully resolved URI for spec.subscriber.
+	SubscriberURI *apis.URL `json:"subscriberURI,omitempty"`
```
the correct json should be `subscriberUri`

n3wscott commented on pkg/apis/messaging/v1beta1/subscription_types.go:
```diff
@@ -0,0 +1,147 @@
+	SubscriberURI *apis.URL `json:"subscriberURI,omitempty"`
+
+	// ReplyURI is the fully resolved URI for the spec.reply.
+	ReplyURI *apis.URL `json:"replyURI,omitempty"`
```
the correct json should be `replyUri`

n3wscott commented on pkg/apis/messaging/v1beta1/subscription_types.go:
```diff
@@ -0,0 +1,147 @@
+	ReplyURI *apis.URL `json:"replyURI,omitempty"`
+
+	// ReplyURI is the fully resolved URI for the spec.delivery.deadLetterSink.
+	DeadLetterSinkURI *apis.URL `json:"deadLetterSinkURI,omitempty"`
```
the correct json should be deadLetterSinkUri`

Why is the channel `error` and this `deadLetter`?

n3wscott commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -0,0 +1,195 @@
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+	// +optional
+	SubscriberURI *apis.URL `json:"subscriberURI,omitempty"`
```
the correct json is `subscriberUri`

n3wscott commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -0,0 +1,195 @@
+	// +optional
+	SubscriberURI *apis.URL `json:"subscriberURI,omitempty"`
+	// +optional
+	ReplyURI *apis.URL `json:"replyURI,omitempty"`
```
the correct json is `replyUri`

n3wscott commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -0,0 +1,195 @@
+	// +optional
+	ReplyURI *apis.URL `json:"replyURI,omitempty"`
+	// +optional
+	DeadLetterSinkURI *apis.URL `json:"deadLetterSink,omitempty"`
```
this should match the other URIs in the form `deadLetterSinkUri`

vaikas commented on pkg/apis/duck/v1beta1/channelable_types.go:
```diff
@@ -0,0 +1,154 @@
+
+	// DeliverySpec contains options controlling the event delivery
+	// +optional
+	Delivery *eventingduckv1alpha1.DeliverySpec `json:"delivery,omitempty"`
```
My thought was that the delivery spec is new enough that I felt we should not promote that to v1beta1. I tried to put that in the description (easy to miss, I do it all the time :) )

vaikas replied:
@lionelvillard Thoughts on promoting the delivery spec to v1beta1?

n3wscott replied:
I think lemonade had the principle of never promoting, always backport. So v1alpha1 can have v1beta1 but not the other way around and this import breaks that. It also means we are promoting the Delivery api to v1beta1 indirectly because of this import.

vaikas replied:
This has been now hoisted.

vaikas commented on pkg/apis/duck/v1beta1/channelable_types.go:
```diff
@@ -0,0 +1,154 @@
+	SubscribableTypeStatus `json:",inline"`
+	// ErrorChannel is set by the channel when it supports native error handling via a channel
+	// +optional
+	ErrorChannel *corev1.ObjectReference `json:"errorChannel,omitempty"`
```
when you say ref, you mean pointer? Or something else? If a Channel does not support (and none are provided via spec, it could be missing)? And furthermore, are you saying this should be a apis.URL?


Fine with the name change.

n3wscott replied:
Yeah, I think Knative does object refs in specs, and resolved uris in status. BUT it depends on if this extra channel is being created for the channel and not given to the channel, then this might make sense.

So if the goal here is for the parent channel to advertise that it has made another channel for you to subscribe to errors on, then an object ref makes sense and I would just call it a `DeadLetterChannel` and keep it an object ref.

n3wscott replied:
`DeadLetterChannel` for messages that failed to be delivered. 

In the future I could also see `ErrorChannel` or `ErrorSink` as general errors that the channel would like to emit but those might not be because of incoming messages (like they are messages generated from the control plane).

vaikas replied:
ack :)

vaikas commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -0,0 +1,195 @@
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// Status of the subscriber.
+	// +optional
+	Ready corev1.ConditionStatus `json:"ready,omitempty"`
```
actually, it should not be optional. it should be ConditionStatus = [Ready,False,Unknown]

vaikas commented on pkg/apis/messaging/v1beta1/register.go:
```diff
@@ -0,0 +1,57 @@
+// Adds the list of known types to Scheme.
+func addKnownTypes(scheme *runtime.Scheme) error {
+	scheme.AddKnownTypes(SchemeGroupVersion,
+		//		&InMemoryChannel{},
```
I think they are still in there?

vaikas commented on pkg/apis/messaging/v1beta1/register.go:
```diff
@@ -0,0 +1,57 @@
+		//		&InMemoryChannelList{},
+		&Subscription{},
+		&SubscriptionList{},
+		//		&Channel{},
```
yep, thanks, done.

vaikas commented on pkg/apis/messaging/v1beta1/subscription_types.go:
```diff
@@ -0,0 +1,147 @@
+
+	// Delivery configuration
+	// +optional
+	Delivery *eventingduckv1alpha1.DeliverySpec `json:"delivery,omitempty"`
```
see the previous comment.

vaikas commented:
Thanks much! So far I think I addressed all the items that I could that I think we can do without some further input. Here's what I see the big ticket items that need resolving. Pulling them here
so they don't get lost in the comments:
1. Can Subscription specify DLC
2. Should DeliverySpec either be:
- Left out (don't want to promote it to v1beta1?)
- Moved to v1beta1
3. Change Subscription.Status.PhysicalSubscription to ResolvedSubscription
- AND: rename json field: physicalSubscription -> resolved
4. What does (ties in with 1 above) Status expose.

lionelvillard commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -0,0 +1,195 @@
+	// +optional
+	ReplyURI *apis.URL `json:"replyURI,omitempty"`
+	// +optional
+	DeadLetterSinkURI *apis.URL `json:"deadLetterSink,omitempty"`
```
yes to allow fine-grained delivery configuration.

vaikas replied:
Could we clarify the semantics little more? Does the DLC apply to Subscribe/Reply, both? Are there any overlap / issues with specifying these both at the Subscription and Channel level?
Perhaps we should take a look at adding these into the design (which didn't seem to talk about Subscription having a DLC). 
https://docs.google.com/document/d/1qRrzGoHJQO-oc5p-yRK8IRfugd-FM_PXyM7lN5kcqks/edit#heading=h.d08lklgnphgg
@lionelvillard

vaikas commented:
/test pull-knative-eventing-integration-tests

matzew commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -0,0 +1,177 @@
+	// +optional
+	SubscriberURI *apis.URL `json:"subscriberUri,omitempty"`
+	// +optional
+	ReplyURI *apis.URL `json:"replyUri,omitempty"`
```
`replyURI` ?

vaikas replied:
Nope, these got changed to these. These are the cases we should be using for the JSON names. @n3wscott

matzew commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -0,0 +1,177 @@
+	// +optional
+	ReplyURI *apis.URL `json:"replyUri,omitempty"`
+	// +optional
+	DeadLetterSinkURI *apis.URL `json:"deadLetterSinkUri,omitempty"`
```
`deadLetterSinkURI` ?

matzew commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -0,0 +1,177 @@
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+	// +optional
+	SubscriberURI *apis.URL `json:"subscriberUri,omitempty"`
```
`subscriberURI` ?

e.g. https://github.com/knative/eventing/blob/master/pkg/apis/duck/v1alpha1/subscribable_types.go#L55

n3wscott replied:
JSON should be camel-cased in the `subscriberUri ` form. Golang should be `subscriberURI`. https://github.com/knative/eventing/blob/master/pkg/apis/duck/v1alpha1/subscribable_types.go#L55 is wrong.

matzew commented on pkg/apis/messaging/v1beta1/subscription_types.go:
```diff
@@ -0,0 +1,147 @@
+	ReplyURI *apis.URL `json:"replyUri,omitempty"`
+
+	// ReplyURI is the fully resolved URI for the spec.delivery.deadLetterSink.
+	DeadLetterSinkURI *apis.URL `json:"deadLetterSinkUri,omitempty"`
```
URI ?

lionelvillard replied:
I think the convention is `Uri` in the JSON.

lionelvillard commented on pkg/apis/duck/v1beta1/delivery_types.go:
```diff
@@ -0,0 +1,67 @@
+	// DeadLetterSink is the sink receiving event that couldn't be sent to
+	// a destination.
+	// +optional
+	DeadLetterSink *duckv1.Destination `json:"deadLetterSink,omitempty"`
```
should be `deadLetterSinkUri`

vaikas replied:
It's the Destination, so I don't think it actually should be an uri here?

lionelvillard commented on pkg/apis/duck/v1beta1/delivery_types.go:
```diff
@@ -0,0 +1,67 @@
+	// DeadLetterChannel is the reference to the native, platform specific channel
+	// where failed events are sent to.
+	// +optional
+	DeadLetterChannel *corev1.ObjectReference `json:"deadLetterChannel,omitempty"`
```
maybe the type should be api.URL.

vaikas replied:
In the status, it should be a reference to a Channel (well, really, Subscribable). That way you can Subscribe to the events that have failed. Make sense?

n3wscott replied:
I had the same thought until I realized that this really is a ref to a channel that the owning channel created. I am +1 on DeadLetterChannel

n3wscott replied:
Maybe this object also needs a deadlettersinkuri? If the resource has a delivery spec in the spec there is currently no place to show a resolved delivery deadleattersink.

Or there are two kinds of status objects, one where the channel makes you a deadletter channel, and ones that use the resolved deadletter destination

lionelvillard commented on pkg/apis/messaging/v1beta1/subscription_types.go:
```diff
@@ -0,0 +1,147 @@
+	duckv1.Status `json:",inline"`
+
+	// PhysicalSubscription is the fully resolved values that this Subscription represents.
+	PhysicalSubscription SubscriptionStatusPhysicalSubscription `json:"physicalSubscription,omitempty"`
```
what about inlining it? No more physicalSubscription. serving does this a lot.

n3wscott replied:
Explain more, how would it look in this API? Sounds interesting!

lionelvillard commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -0,0 +1,177 @@
+
+// SubscribableTypeStatus shows how we expect folks to embed Subscribable in their Status field.
+type SubscribableTypeStatus struct {
+	SubscribableStatus *SubscribableStatus `json:"subscribableStatus,omitempty"`
```
`...Status` feels redundant:
```yaml
status:
  subscribableStatus: 
    subscribers:  ...
```

vaikas replied:
ack, cleaned this up + channelable stuff as well.

lionelvillard commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -0,0 +1,177 @@
+
+// SubscribableTypeSpec shows how we expect folks to embed Subscribable in their Spec field.
+type SubscribableTypeSpec struct {
+	Subscribable *Subscribable `json:"subscribable,omitempty"`
```
I always felt that `subscribable` is redundant. Why not just:
```yaml
spec:
  subscribers: 
  - ...
```

vaikas replied:
ack :)

mattmoor commented on pkg/apis/messaging/v1beta1/subscribable_channelable_validation_test.go:
```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/messaging/v1beta1/subscription_defaults_test.go:
```diff
@@ -0,0 +1,28 @@
+/*
+Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/messaging/v1beta1/subscription_lifecycle_test.go:
```diff
@@ -0,0 +1,267 @@
+/*
+Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/messaging/v1beta1/subscription_validation_test.go:
```diff
@@ -0,0 +1,440 @@
+/*
+Copyright 2018 The Knative Authors. All Rights Reserved.
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors. All Rights Reserved.
```

n3wscott commented on pkg/apis/duck/v1beta1/channelable_types.go:
```diff
@@ -0,0 +1,152 @@
+
+	// DeliverySpec contains options controlling the event delivery
+	// +optional
+	Delivery *DeliverySpec `json:"delivery,omitempty"`
```
Should Delivery be inside of Subscribable?

vaikas replied:
So, currently the SubscriberSpec has a DeadLetterSinkURI in there, which is apis.URL. I wonder if that should be replaced with a Delivery? Otherwise, I don't see how the retry policy can with the current semantics be wired through to the Subscribable? So, that seems to mean we should put it in SubscriberSpec, thoughts?
Basically SubscriberSpec today has:
	// +optional
	DeadLetterSinkURI *apis.URL `json:"deadLetterSinkUri,omitempty"`

Seems that should be replaced with:
     Delivery *DeliverySpec `json:"deliver, omitempty"`

@lionelvillard @n3wscott

n3wscott replied:
I like this idea.

vaikas replied:
ok, I'll bang it out.

n3wscott commented on pkg/apis/duck/v1beta1/channelable_types.go:
```diff
@@ -0,0 +1,152 @@
+// Channelable is a skeleton type wrapping Subscribable and Addressable in the manner we expect resource writers
+// defining compatible resources to embed it. We will typically use this type to deserialize
+// Channelable ObjectReferences and access their subscription and address data.  This is not a real resource.
+type Channelable struct {
```
If Delivery moved into Subscribable, then we can drop Channelable all together and a channel becomes 

Addressable + Subscribable. 

What do you think?

lionelvillard replied:
make sense to me. Big change though....

vaikas replied:
Besides removing one duck type (which I kind of think adds clarity), are there other benefits to removing it? I think being able to operate on a Channel ducktype as a single object in the consuming code base is kind of nice (instead of operating on Addressable / Subscribable) separately.
Not against it, just not sure what the concrete benefits of it are.

vaikas replied:
sorry, I meant that actually having the Channelable as a separate duck type adds clarity, not removing it. So, what are the benefits for removing it?

n3wscott replied:
Simplify the number of things, channelable does not add anything new, just the union of two other ducks

vaikas replied:
But the code using it now needs to set up two watches if it cares about channels (one for addressable and one for subscribable) and extra serializations / etc.? Am I just confused here? Seems like it has downsides and I'm not seeing the upside of removing the duck.

vaikas replied:
Given the discussion above about Delivery NOT moving to Subscribable (since it needs to be available per Subscription) but moving to SubscriberSpec and the fact that there's also a DeadLetterChannel in the Channel, I don't see this happening. But perhaps I'm missing something here? So it's not only Subscribable, Addressable but also has DeadLetterChannel in the status.

n3wscott commented on pkg/apis/messaging/v1beta1/channel_types.go:
```diff
@@ -0,0 +1,110 @@
+	ChannelTemplate *ChannelTemplateSpec `json:"channelTemplate"`
+
+	// Channel conforms to Duck type Subscribable.
+	Subscribable *eventingduck.Subscribable `json:"subscribable,omitempty"`
```
@vaikas shoudl this move to use the SubscribaleSpec inlined here?

vaikas replied:
yep 👍 Also ChannelStatus should have ChannelableStatus, fixed that also. ✅

n3wscott commented on pkg/apis/messaging/v1beta1/channel_types.go:
```diff
@@ -0,0 +1,110 @@
+}
+
+// GetGroupVersionKind returns GroupVersionKind for Channels.
+func (dc *Channel) GetGroupVersionKind() schema.GroupVersionKind {
```
I think this is suppose to be in *_lifecycle.go

n3wscott commented on pkg/apis/messaging/v1beta1/channel_types.go:
```diff
@@ -0,0 +1,110 @@
+}
+
+// GetUntypedSpec returns the spec of the Channel.
+func (c *Channel) GetUntypedSpec() interface{} {
```
I think this is suppose to be in *_lifecycle.go

n3wscott commented on pkg/apis/messaging/v1beta1/in_memory_channel_types.go:
```diff
@@ -0,0 +1,103 @@
+// arguments for a Channel.
+type InMemoryChannelSpec struct {
+	// Channel conforms to Duck type Subscribable.
+	Subscribable *eventingduckv1beta1.Subscribable `json:"subscribable,omitempty"`
```
This should be inlined .SubscribableSpec ?

n3wscott commented on pkg/apis/messaging/v1beta1/in_memory_channel_types.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// ChannelStatus represents the current state of a Channel.
+type InMemoryChannelStatus struct {
```
How the shape of the channelabe shows up here (meaning it does not use the duck) kinda makes validated that channelable should be removed.

vaikas replied:
I think it's more useful for code that operates on Channelable types, so they don't have to operate on two ducks, they can just use one?

n3wscott commented on pkg/apis/messaging/v1beta1/in_memory_channel_types.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// GetGroupVersionKind returns GroupVersionKind for InMemoryChannels
+func (imc *InMemoryChannel) GetGroupVersionKind() schema.GroupVersionKind {
```
should be in *_lifecyce.go based on serving

n3wscott commented on pkg/apis/messaging/v1beta1/in_memory_channel_types.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// GetUntypedSpec returns the spec of the InMemoryChannel.
+func (i *InMemoryChannel) GetUntypedSpec() interface{} {
```
should be in *_lifecyce.go based on serving

n3wscott commented on pkg/apis/messaging/v1beta1/subscription_defaults.go:
```diff
@@ -0,0 +1,29 @@
+	s.Spec.SetDefaults(ctx)
+}
+
+func (ss *SubscriptionSpec) SetDefaults(ctx context.Context) {
```
We might want to think about what in delivery should now be default for subscription and add it to v1beta1

n3wscott commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -85,69 +68,65 @@ type SubscriberStatus struct {
+	// SubscribableStatus is the part where SubscribableStatus object is
 	// configured as to be compatible with Subscribable contract.
-	Status SubscribableTypeStatus `json:"status"`
+	Status SubscribableStatus `json:"status"`
```
I still want to question `status`. Is it really just a quick way to know `spec.subscriber[x].generation == status.subscriber[x].generation`?

If so, I think it should be removed.

If not, and it is the status of the subscriber, it should be removed because it is not the job of the channel to look up the subscribers.

If it is the status of the data plane result (like, we are getting 2xx's) I also think it should be renamed or rethought because if this is false, it is useless without a reason and message.

vaikas replied:
I was thinking it would be that the subscription has been wired up. You asked for a subscription in the Spec, reconciler then goes and makes that, how would you know if something went wrong? But, I'm reading some of the more of the codes now.

n3wscott replied:
Then maybe you are pro-"add a reason and a message"? 

`status.subscriber[x].ready: false` falls short of being useful info to act on.

vaikas replied:
Ack. @n3wscott and I talked offline. So the spec.Array and status.Array has precedent in things like pod (container in spec -> containerstatuses). so using array in spec and array in status to match part is all good. I'll create a more targeted issue for possibly renaming / modifying the fields within the SubscriberStatus struct.

vaikas replied:
I created this issue to track this in specific. As we move forward with the implementation of the controller and channels, etc. we will revisit this as part of that issue:
https://github.com/knative/eventing/issues/2416

n3wscott commented on pkg/apis/duck/v1beta1/subscribable_types.go:
```diff
@@ -157,21 +136,20 @@ func (c *SubscribableType) Populate() {
 // GetListType implements apis.Listable
-func (c *SubscribableType) GetListType() runtime.Object {
-	return &SubscribableTypeList{}
+func (c *Subscribable) GetListType() runtime.Object {
```
this reads so much better, nice work.

n3wscott commented on pkg/apis/messaging/v1beta1/channel_types.go:
```diff
@@ -66,24 +64,14 @@ type ChannelSpec struct {
-
-	// Subscribers is populated with the statuses of each of the Channelable's subscribers.
-	eventingduck.SubscribableTypeStatus `json:",inline"`
+	// Channel conforms to ChannelableStatus
```
🎉

n3wscott commented on pkg/apis/messaging/v1beta1/in_memory_channel_validation.go:
```diff
@@ -30,13 +30,11 @@ func (imc *InMemoryChannel) Validate(ctx context.Context) *apis.FieldError {
-				fe.Details = "expected at least one of, got none"
-				errs = errs.Also(fe.ViaField(fmt.Sprintf("subscriber[%d]", i)).ViaField("subscribable"))
-			}
+	for i, subscriber := range imcs.SubscribableSpec.Subscribers {
```
this looks much nicer too

vaikas replied:
:)

n3wscott commented:
/lgtm
